Add quirk for TripAdvisor to support mixed content in their Panorama feature

brentfulgham opened the pull request:
#### 0a2730dd7efa7b0fceccec9ae9b3f5d14acebd7f
<pre>
Add quirk for TripAdvisor to support mixed content in their Panorama feature
<a href="https://bugs.webkit.org/show_bug.cgi?id=273425">https://bugs.webkit.org/show_bug.cgi?id=273425</a>
&lt;<a href="https://rdar.apple.com/112851939">rdar://112851939</a>&gt;

Reviewed by Matthew Finkel.

Add a quirk to allow TripAdvisor.com&apos;s Panorama feature to work despite their use of mixed
content in conjunction with crossorigin=&quot;&quot; style. This brings Safari behavior in line with
Chrome, which performs the insecure content upgrade (even though the specification indicates
this should not happen).

* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::destinationIsImageAudioOrVideo):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsRelaxedCorsMixedContentCheckQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/278170@main">https://commits.webkit.org/278170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15d690906a97fc16591f42efaec32b4afc4f4816

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49590 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28875 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52637 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52832 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/266 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51894 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26495 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40481 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51690 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42717 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21596 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23864 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43901 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7961 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45782 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44406 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54409 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20851 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25944 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26788 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7153 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25669 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebCore/loader/MixedContentChecker.cpp:
```diff
@@ -184,11 +190,11 @@ bool MixedContentChecker::shouldUpgradeInsecureContent(LocalFrame& frame, IsUpgr
         || (!shouldUpgradeIPAddressAndLocalhostForTesting && URL::hostIsIPAddress(url.host()))
         // 4.1.4 request’s destination is not "image", "audio", or "video".
-        || (destination != FetchOptions::Destination::Audio && destination != FetchOptions::Destination::Image && destination != FetchOptions::Destination::Video)
+        || (!destinationIsImageAudioOrVideo(destination))
```
Maybe we need to drop this condition, in favor of the one below (and move/adjust the comment)?

brentfulgham replied:
I think we still want this, for cases where the mode is NOT cors. If just use the changes I propose below, we would decide not to check for image/audio/media if the mode was non-cors, and allow he load.

sysrqb commented on Source/WebCore/loader/MixedContentChecker.cpp:
```diff
@@ -184,11 +190,11 @@ bool MixedContentChecker::shouldUpgradeInsecureContent(LocalFrame& frame, IsUpgr
         || (destination == FetchOptions::Destination::Image && initiator == Initiator::Imageset)
         // and CORS is excluded
-        || mode == FetchOptions::Mode::Cors)
+        || (mode == FetchOptions::Mode::Cors && (!document->quirks().needsRelaxedCorsMixedContentCheckQuirk() || !destinationIsImageAudioOrVideo(destination))))
```
Or we could just remove this because we're essentially ignoring the fetch mode and we shouldn't reach this condition if we already checked `!destinationIsImageAudioOrVideo` and that was true. Maybe just add a FIXME and we should bring the up with WebAppSec.

brentfulgham replied:
I think we do want this (unless the Fetch mode is impractical given the state of the web today). My reasoning is that some sites have Fetch::mode set properly and want it enforced, and we should honor it.

This change just says:
"If a site has Cors set, if its one of the (currently one) site(s) that break if we block the load as the specification requires for image/audio/media, allow the upgrade. Otherwise, do what the site author wanted."

Another possibility would be for me to pass the Fetch Mode to the Quirk for checking, but that would involve telling Quirks.cpp about FetchOptions, which just adds more include complexity. So I thought this check might be simpler.

brentfulgham replied:
@sysrqb I revised line 197 slightly to make it clearer that we are checking:
(1) Fetch mode is one where we are supposed to reject upgrading
(2) -AND- we are NOT in the special state of:
(a) We are on a site needing the Quirk
(b) The load is also for image/audio/media

brentfulgham committed:
Add quirk for TripAdvisor to support mixed content in their Panorama feature
https://bugs.webkit.org/show_bug.cgi?id=273425
<rdar://112851939>

Reviewed by Matthew Finkel.

Add a quirk to allow TripAdvisor.com's Panorama feature to work despite their use of mixed
content in conjunction with crossorigin="" style. This brings Safari behavior in line with
Chrome, which performs the insecure content upgrade (even though the specification indicates
this should not happen).

* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::destinationIsImageAudioOrVideo):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsRelaxedCorsMixedContentCheckQuirk const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/278170@main
